File: panel/models/interaction_store.py
"""
Defines the InteractionStore model which provides a centralized store
for cross-component interaction events (hover, selection, viewport, row selection).
"""
from __future__ import annotations

import typing as t

from bokeh.core.enums import enumeration
from bokeh.core.properties import (
    Any, Bool, Dict, Int, List, String,
)
from bokeh.events import ModelEvent
from bokeh.models import Model

from ..config import config
from ..io.resources import bundled_files
from ..util import classproperty

InteractionEventKind = t.Literal["hover", "selection", "viewport", "row_selection"]

InteractionEventKinds = enumeration("hover", "selection", "viewport", "row_selection")


class InteractionEvent(ModelEvent):

    event_name = 'interaction_event'

    def __init__(
        self,
        model,
        kind: InteractionEventKind | None = None,
        source: str | None = None,
        source_id: str | None = None,
        payload: t.Any = None,
        selection: t.Any = None,
        viewport: t.Any = None,
    ):
        self.kind = kind
        self.source = source
        self.source_id = source_id
        self.payload = payload
        self.selection = selection
        self.viewport = viewport
        super().__init__(model=model)

    def __repr__(self):
        return (
            f'{type(self).__name__}(kind={self.kind!r}, source={self.source!r}, '
            f'source_id={self.source_id!r}, selection={self.selection!r}, '
            f'viewport={self.viewport!r})'
        )


class InteractionStore(Model):
    """
    A centralized store for collecting and distributing interaction events
    across multiple Panel components (Plotly, Vega, ECharts, Tabulator, etc.).

    Components publish their events to the store, and other components or
    Python callbacks can subscribe to receive filtered events by kind
    or source.

    Each event has a normalized schema:
        - kind: "hover" | "selection" | "viewport" | "row_selection"
        - source: human-readable source name (e.g. "plotly", "vega")
        - source_id: model id of the originating component
        - timestamp: monotonic timestamp in ms
        - payload: original raw event from the library
        - selection: normalized selection data (when applicable)
            - mode: "point" | "range" | "rows"
            - indices: list of integer indices
            - values: list of {field: value} dicts
            - ranges: (mode="range" only) {axis: [min, max]}
        - viewport: normalized viewport data (when applicable)
            - ranges: {axis: [min, max]}
    """

    events = List(
        Dict(String, Any),
        default=[],
        help="""
        List of normalized interaction events. See class docstring for schema.
        """,
    )

    hover_events = List(
        Dict(String, Any),
        default=[],
        help="Filtered view: events of kind 'hover'",
    )

    selection_events = List(
        Dict(String, Any),
        default=[],
        help="Filtered view: events of kind 'selection'",
    )

    viewport_events = List(
        Dict(String, Any),
        default=[],
        help="Filtered view: events of kind 'viewport'",
    )

    row_selection_events = List(
        Dict(String, Any),
        default=[],
        help="Filtered view: events of kind 'row_selection'",
    )

    sources = Dict(
        String,
        String,
        default={},
        help="""
        Mapping from source model id to a human-readable source name.
        Components register themselves here when publishing their first event.
        """,
    )

    max_history = Int(
        default=100,
        help="Maximum number of events to keep in the history.",
    )

    _autoclear = Bool(
        default=False,
        help="If True, automatically clear events after they have been dispatched.",
    )

    __javascript_raw__: list[str] = []

    @classproperty
    def __javascript__(cls):
        return bundled_files(cls)

    @classproperty
    def __js_skip__(cls):
        return {}

File: panel/models/interaction_store.ts
import type {Attrs} from "@bokehjs/core/types"
import type * as p from "@bokehjs/core/properties"
import {ModelEvent} from "@bokehjs/core/bokeh_events"
import {Model} from "@bokehjs/model"
import {View} from "@bokehjs/core/view"

export type InteractionEventKind = "hover" | "selection" | "viewport" | "row_selection"

export interface InteractionSelectionPoint {
  mode: "point"
  indices: number[]
  values: Array<{[field: string]: any}>
}

export interface InteractionSelectionRange {
  mode: "range"
  ranges: {[axis: string]: [any, any]}
  indices?: number[]
  values?: Array<{[field: string]: any}>
}

export interface InteractionSelectionRows {
  mode: "rows"
  indices: number[]
  values: Array<{[field: string]: any}>
}

export type InteractionSelection =
  | InteractionSelectionPoint
  | InteractionSelectionRange
  | InteractionSelectionRows

export interface InteractionViewport {
  ranges: {[axis: string]: [any, any]}
}

export interface InteractionEventPayload {
  kind: InteractionEventKind
  source: string
  source_id: string
  timestamp: number
  payload: any
  selection?: InteractionSelection
  viewport?: InteractionViewport
}

export class InteractionEvent extends ModelEvent {
  constructor(
    readonly kind: InteractionEventKind,
    readonly source: string,
    readonly source_id: string,
    readonly payload: any,
    readonly selection?: InteractionSelection,
    readonly viewport?: InteractionViewport,
  ) {
    super()
  }

  protected override get event_values(): Attrs {
    return {
      model: this.origin,
      kind: this.kind,
      source: this.source,
      source_id: this.source_id,
      payload: this.payload,
      selection: this.selection,
      viewport: this.viewport,
    }
  }

  static {
    this.prototype.event_name = "interaction_event"
  }
}

export class InteractionStoreView extends View {
  declare model: InteractionStore
}

export namespace InteractionStore {
  export type Attrs = p.AttrsOf<Props>

  export type Props = Model.Props & {
    events: p.Property<InteractionEventPayload[]>
    hover_events: p.Property<InteractionEventPayload[]>
    selection_events: p.Property<InteractionEventPayload[]>
    viewport_events: p.Property<InteractionEventPayload[]>
    row_selection_events: p.Property<InteractionEventPayload[]>
    sources: p.Property<{[key: string]: string}>
    max_history: p.Property<number>
    _autoclear: p.Property<boolean>
  }
}

export interface InteractionStore extends InteractionStore.Attrs {}

export class InteractionStore extends Model {
  declare properties: InteractionStore.Props

  constructor(attrs?: Partial<InteractionStore.Attrs>) {
    super(attrs)
  }

  static override __module__ = "panel.models.interaction_store"

  static {
    this.prototype.default_view = InteractionStoreView

    this.define<InteractionStore.Props>(({Any, Bool, Int, List, Str}) => ({
      events: [ List(Any), [] ],
      hover_events: [ List(Any), [] ],
      selection_events: [ List(Any), [] ],
      viewport_events: [ List(Any), [] ],
      row_selection_events: [ List(Any), [] ],
      sources: [ Any, {} ],
      max_history: [ Int, 100 ],
      _autoclear: [ Bool, false ],
    }))
  }

  publish_event(event: Omit<InteractionEventPayload, "timestamp">): InteractionEventPayload {
    const full: InteractionEventPayload = {
      ...event,
      timestamp: Date.now(),
    }

    const events = [...this.events, full]
    const overflow = events.length - this.max_history
    const trimmed = overflow > 0 ? events.slice(overflow) : events

    this.events = trimmed
    this._update_filtered_views(trimmed)

    if (event.source != null && this.sources[event.source_id] == null) {
      this.sources = {...this.sources, [event.source_id]: event.source}
    }

    this.trigger_event(new InteractionEvent(
      full.kind, full.source, full.source_id,
      full.payload, full.selection, full.viewport,
    ))

    if (this._autoclear) {
      setTimeout(() => this.clear(full.kind, full.source_id), 0)
    }

    return full
  }

  subscribe(
    callback: (event: InteractionEventPayload) => void,
    kind_filter?: InteractionEventKind | InteractionEventKind[],
    source_filter?: string | string[],
  ): () => void {
    const kinds = kind_filter == null
      ? null
      : Array.isArray(kind_filter) ? new Set(kind_filter) : new Set([kind_filter])
    const sources = source_filter == null
      ? null
      : Array.isArray(source_filter) ? new Set(source_filter) : new Set([source_filter])

    const handler = () => {
      const latest = this.events[this.events.length - 1]
      if (latest == null) {
        return
      }
      if (kinds != null && !kinds.has(latest.kind)) {
        return
      }
      if (sources != null && !sources.has(latest.source_id)) {
        return
      }
      callback(latest)
    }

    this.on_change(this.properties.events, handler)

    return () => {
      this.remove_on_change(this.properties.events, handler)
    }
  }

  clear(kind?: InteractionEventKind, source_id?: string): void {
    let filtered = this.events
    if (kind != null) {
      filtered = filtered.filter((e) => e.kind !== kind)
    }
    if (source_id != null) {
      filtered = filtered.filter((e) => e.source_id !== source_id)
    }
    this.events = filtered
    this._update_filtered_views(filtered)
  }

  clear_all(): void {
    this.events = []
    this._update_filtered_views([])
  }

  get_events(
    kind?: InteractionEventKind,
    source_id?: string,
  ): InteractionEventPayload[] {
    return this.events.filter((e) => {
      if (kind != null && e.kind !== kind) {
        return false
      }
      if (source_id != null && e.source_id !== source_id) {
        return false
      }
      return true
    })
  }

  _update_filtered_views(events: InteractionEventPayload[]): void {
    const hover: InteractionEventPayload[] = []
    const selection: InteractionEventPayload[] = []
    const viewport: InteractionEventPayload[] = []
    const row_selection: InteractionEventPayload[] = []

    for (const e of events) {
      switch (e.kind) {
        case "hover":
          hover.push(e)
          break
        case "selection":
          selection.push(e)
          break
        case "viewport":
          viewport.push(e)
          break
        case "row_selection":
          row_selection.push(e)
          break
      }
    }

    this.hover_events = hover
    this.selection_events = selection
    this.viewport_events = viewport
    this.row_selection_events = row_selection
  }
}
